### ISSUE TITLE
Cudnn header files should be copied into build package as well 

### ISSUE THREAD

**Issue by mszhanyi** [pytorch/pytorch#47743]
## 🚀 Feature
Cudnn headers should be copied to build package as well.

## Motivation
Logically, we have copied the cudnn dlls to build package, the head files should be copied as well.
In reality, If the user is using C++ frontend, he/she might need to set CUDNN_INCLUDE_DIR in CMAKE, It'd better to keep the head files consistent with the dlls.

cc @peterjc123 @maxluk @nbcsm @guyang3532 @gunandrose4u @smartcat2010 @mszhanyi

**Comment by peterjc123** [comment]
Why? The DLL is not included for the development purposes. It just serves as a dependency for `torch_cuda.dll`. And btw, it is not enough to pack in only the cudnn header for the development if he really uses something directly from cudnn. We need the static library for linking, right? I'm quite uncertain how this could help our users.

**Comment by mszhanyi** [comment]
If we want to build one demo application as https://tutorials.pytorch.kr/advanced/cpp_frontend.html in windows with cuda.
it would throw exception like below when we run `cmake -DCMAKE_PREFIX_PATH=/path/to/libtorch`
> Your installed Caffe2 version uses cuDNN but I cannot find the cuDNN
>   libraries.  Please set the proper cuDNN prefixes and / or install cuDNN.
One solution is to add CUDNN_INCLUDE_PATH and CUDNN_LIBRARY_PATH into the CMakeLists.txt.
If those header files in CUDNN_INCLUDE_PATH aren't match the copied cudnn dlls, it would throw another exception that cudnn version isn't match

**Comment by peterjc123** [comment]
@ezyang Is CUDNN a public dependency of LibTorch? If it is not, then maybe we should stop finding it during cmake configuration.

**Comment by peterjc123** [comment]
> One solution is to set CUDNN_INCLUDE_PATH and CUDNN_LIBRARY_PATH into the CMakeLists.txt.
> If those header files in CUDNN_INCLUDE_PATH aren't match the copied cudnn dlls, it would throw another exception that cudnn version isn't match

Sorry, I was lost in your description. Are you describing a problem or a solution?

**Comment by mszhanyi** [comment]
> > One solution is to set CUDNN_INCLUDE_PATH and CUDNN_LIBRARY_PATH into the CMakeLists.txt.
> > If those header files in CUDNN_INCLUDE_PATH aren't match the copied cudnn dlls, it would throw another exception that cudnn version isn't match
> 
> Sorry, I was lost in your description. Are you describing a problem or a solution?

Sorry, I didn't make it very clear.
set CUDNN_INCLUDE_PATH can solve the exception that cuDNN isn't found.

But it would be a problem how to set CUDNN_INCLUDE_PATH. Now, we couldn't use the path of downloaded lib_torch. Users may have to install the cudnn by themself.

**Comment by mszhanyi** [comment]
@peterjc123 , any more comments?

**Comment by ezyang** [comment]
> Is CUDNN a public dependency of LibTorch? If it is not, then maybe we should stop finding it during cmake configuration.

In Dependencies.cmake:

```
    if(CAFFE2_USE_CUDNN)
      list(APPEND Caffe2_PUBLIC_CUDA_DEPENDENCY_LIBS caffe2::cudnn)
    else()
```

But I couldn't find any reference to cudnn in cmake/TorchConfig.cmake.in, so it seems quite possible that you don't get cudnn dep by default. But the best way to figure this out is to try configuring TorchConfig cmake and see what you get; I always find it hard to figure out what the runtime behavior of cmake is just by reading. It seems like it would be good to figure this out first.

**Comment by Luthaf** [comment]
This is an issue for me on Linux as well. I'm trying to build a C++ extension to PyTorch, using CMake as the build system. I run `find_package(Torch)` in my code, and this ends up trying to find `Caffe2` and in turn `CUDNN`. I'm trying to build this extension to use it from Python, so I'm relying on the versions of the libraries provided with the different Python packages.

The torch package on PyPI depends on the `nvidia-cudnn-cu11` package, which provides both library and headers, so everything is fine there. But the `pytorch` package on conda bundles its own `libcudnn.so.8`, and does not provide the header, making the cmake search for CUDNN fail. 

I would be willing to contribute something to ensure the headers are also bundled with the shared library if you think this is the best way to solve this issue.

**Comment by ezyang** [comment]
cc @malfet

**Comment by malfet** [comment]
Wow, that's an old issue. IMO we should not be bundling `cudnn.h` as neither libtorch nor extension cmake projects really needs those.
Since PyTorch-1.10 cudnn should be internal dependency of libtorch cuda and as such cmake should not need this.
If this is not the case, please create a new issue and it should be investigated on priority.

**Comment by malfet** [comment]
> OK. I will accept a patch that bundles the headers, but ONLY if we are bundling the shared library itself. This would have to modify our conda packaging in https://github.com/pytorch/builder I haven't the foggiest how to do it, might be hard.

@ezyang please don't, as neither extension nor libtorch developers should really be needing this header.
If somebody really wants to do this, then it would also make sense to include whole set of [`cudnn_frontend`](https://github.com/NVIDIA/cudnn-frontend) headers, which is currently a default interface for integrating CuDNN into libtorch_cuda

**Comment by Luthaf** [comment]
I agree that I don't need this header, but `find_package(Torch)` is trying to find them. I would also be fine with changes on the CMake side to remove the check for CuDNN here https://github.com/pytorch/pytorch/blob/5b2ab0dd4f9b75657ba0e83641cd91864b4f23d1/cmake/public/cuda.cmake#L246

or make this check not fail if it can not find the CuDNN headers (i.e. making CuDNN a private dependency of Caffe2 instead of a public one)

**Comment by ezyang** [comment]
I saw you submitted us the bundling patch. Do you think the directly fix find_package fix is feasible? It's hard for me to tell on a cursory look.

**Comment by Luthaf** [comment]
Yes, I went with the bundling since I'm not 100% sure I understand the build system of PyTorch. I can try to give it a look when I have some more time. Do you know if there is some documentation for the build system/code organization somewhere?

Caffe2 main repository is this one, right? And there is no Caffe2 library, it always gets statically linked inside libtorch?

**Comment by LaurentMazare** [comment]
@Luthaf  I see that the PR bundles the CuDNN headers only, do you think it would be possible to bundle more cuda headers so that including `cuda_runtime_api.h` works (and so maybe all cuda/cudnn headers)?
The main use case for me (mentioned this in #55454 ) would be being able to compile C++ code that uses cuda graphs or cuda streams, to do so you would want to include `ATen/cuda/CUDAGraph.h` which has a dependency on `c10/cuda/CUDAStream.h` which has a dependency on `cuda_runtime_api.h` (maybe another way would be to remove this dependency but I imagine it's likely to be a larger change, and it would be nice for the headers to be "self-contained" anyway).

**Comment by Luthaf** [comment]
I'm unfamiliar with this part of CUDA. Does the pytorch distribution also includes the shared/static libraries for CUDA runtime? Or are these supposed to be installed from another location but `c10/cuda/CUDAStream.h` is unable to find them when compiling `tch-rs`?

**Comment by anthonyalayo** [comment]
I wanted to revive this chat, @Luthaf @LaurentMazare @malfet.

I could be going about this wrong way, but here's what I did in my build:
1. Run on an x64 machine with CUDA installed
2. Use `find_package(CUDAToolkit)` to get its headers/libs included
3. Use `find_package(Torch)` to get its headers/libs included

What happens after that is the following:
```
ldd binary 
...
        libtorch.so => /home/test/cmake-build-debug/pytorch/torch/lib/libtorch.so (0x00007f7ddf2eb000)
...
        libcudart.so.12 => /usr/local/cuda-12.1/targets/x86_64-linux/lib/libcudart.so.12 (0x00007f7d80210000)
        libcudart-9335f6a2.so.12 => /home/test/cmake-build-debug/pytorch/torch/lib/libcudart-9335f6a2.so.12 (0x00007f7d7f431000)
...
``` 

There's 2 instances of cuda's RT library -- the one installed on my machine, and the one that came out of `find_package(Torch)`. If we package the cuda RT library, why not also package the headers? 

Otherwise, what's the best way to deal with this?

**Comment by malfet** [comment]
@anthonyalayo can you please explain, how inclusion of cudnn headers will help with two instances of cuda runtime?

Considering that we are bundling cuda dynamic libraries with PyTorch, we should reconsider the strategy of mangling just cudart name, but this requires a new issue. @atalman mind looking into that?

**Comment by anthonyalayo** [comment]
> @anthonyalayo can you please explain, how inclusion of cudnn headers will help with two instances of cuda runtime?

@malfet I agree this and that are different, but they all lumped into the general problem of adding the libs without the includes. Let me step back a second and ask, what's the expected developer experience with PyTorch?

**One angle is:**
- Use the docker images off of https://hub.docker.com/r/pytorch/pytorch/
- Use the `devel` image for your build stage, and `runtime` image for your run stage
- Using those, the headers (for eg. CUDNN headers), will be available:
https://hub.docker.com/layers/pytorch/pytorch/2.1.0-cuda12.1-cudnn8-devel/images/sha256-fe174e1e257d29976c99ebe9832d9bb20bd9706ea8eff1482cc9af261998c48d?context=explore

```
RUN |1 TARGETARCH=amd64 /bin/sh -c apt-get update && apt-get install -y --no-install-recommends     ${NV_CUDNN_PACKAGE}     ${NV_CUDNN_PACKAGE_DEV}     && apt-mark hold ${NV_CUDNN_PACKAGE_NAME}     && rm -rf /var/lib/apt/lists/* # buildkit
```

However there's a couple caveats to that flow that actually hit my use case:

1. There are only amd64 builds (I have build configurations with amd64 and arm64)
2. The images are quite big, most likely due to the inclusion of PyTorch itself. However if you have PyTorch as part of your own CMake build flow, there are optimizations you can do to make this smaller (ie. mount caches)
3. We have development off of Docker as well, and in that case having a script to download PyTorch's builds is a better option. It gives a unified experience with and without Docker.

**What I did was:**
- Grab an image off of https://hub.docker.com/r/nvidia/cuda
- It has `base`, `runtime`, and `devel` images similar to pytorch/pytorch, but they are significantly smaller in size:
> - base: Includes the CUDA runtime (cudart)
> - runtime: Builds on the base and includes the [CUDA math libraries](https://developer.nvidia.com/gpu-accelerated-libraries), and [NCCL](https://developer.nvidia.com/nccl). A runtime image that also includes [cuDNN](https://developer.nvidia.com/cudnn) is available.
> - devel: Builds on the runtime and includes headers, development tools for building CUDA images. These images are particularly useful for multi-stage builds.

- Write a small script to download a copy of PyTorch. In this case, download the CUDA version of it.
- Build and link the project using PyTorch and CUDA. 

Where the last step being what I commented earlier:

> Use find_package(CUDAToolkit) to get its headers/libs included
> Use find_package(Torch) to get its headers/libs included

But that raises a lot of questions that are relevant to this discussion. If one was to use the nvidia/cuda images, which provides all the support you need minus PyTorch itself, what should one do here? You end up with linking multiple libraries. 

This is because:
1. The CUDA libraries have unique names in the PyTorch release (ie. libcudart-d0da41ae.so.11.0)
2. The CUDA libraries get pulled in when calling `find_package(Torch)`
3. You need the headers provided by the nvidia/cuda images

If the headers were also included with PyTorch, then one could simply take any clean OS base, download a copy of PyTorch, and be able to build without any issues. Why include the libraries, but not the headers?

**Comment by malfet** [comment]
@anthonyalayo do you mind creating a separate issue or post to https://dev-discuss.pytorch.org/ to elaborate on your development pains? 
IMO PyTorch binary distributions pursue following goals:
 - Running `pip install torch` should install workable version of PyTorch on clean OS image that only has Python runtime (similarly for `conda install` + some caveats about `torch.compile` support)
 - PyTorch CUDA builds are not alternative disributions of cuda toolkit, i.e. one should not be needing to have CUDA toolkit nor NVCC installed to be able to build C++ application that links against libtorch and creates a tensor on GPU
 - Following logic outlined about building CUDA extensions should require one to have CUDA toolkit installed on the system.

So, if you saying that you need CuDNN header just to compile something like:
```cpp
#include <iostream>
#include <torch/torch.h>
int main() {
  std::cout <<  torch::eye(3) << std::endl;
}
```
then this needs to be fixed, but including CuDNN headers feels like a wrong fix for the problem, as `torch.h` should not include any of those.

**Comment by anthonyalayo** [comment]
That's fair @malfet, I forgot to include an important point above. If one _also_ wants to use CUDA headers in the same PyTorch program like the following:

```
#include <cuda_runtime.h>

...
cudaGetDeviceProperties( &deviceProp, 0 );
```

That's when the issues above start to occur.

**Comment by malfet** [comment]
@anthonyalayo sure, if you want to make an application that makes  CUDA API calls you need to have CUDA toolkit installed, not sure why having PyTorch installed should change this. 
I accept the argument, that pytorch -devel docker image perhaps should cuda toolkit, but I don't agree that this the same should stay true for say PyTorch wheels or libtorch packages. 

To give another example, installing `python3-dev` package on Ubuntu gives one access to Python public API, but it does not include all the headers that cpython depends on during its build time. Similarly `libtorch` should not have to expose all of it internal dependencies, only headers necessary to access its public APIs

**Comment by anthonyalayo** [comment]
@malfet agreed, the complication comes with modern CMake. The following is what causes the conflict:

> Use find_package(CUDAToolkit) to get its headers/libs included
> Use find_package(Torch) to get its headers/libs included

I guess a reverse of the question would be -- Is it possible to grab a built PyTorch CUDA flavor (ie. 118 or 121) _without_ the libraries included? Does that break anything?

**Comment by taras-janea** [comment]
Closing this issue due to inactivity. If you have any related concerns, please don't hesitate to open a new issue.
